### PR TITLE
Add `Op` corresponding to `scipy.linalg.solve_discrete_are`

### PR DESCRIPTION
**Pull request by jessegrabowski**
### Motivation for these changes
Move `solve_discrete_are` from `pymc_experimental/statespace/pytensor_scipy.py` to here.

I also did a few cleanups in `tensor.slinalg.py`:

- Removed unused helper function `is_complex_obj`
- Removed unused `solvetriangular` instance of `SolveTriangular` (the `SolveTriangular` `Op` is instead accessed through the `solve_triangular` wrapper function)
- Changed return type hint on `Op` wrapper functions from `TensorVariable` to `Variable` to appease the linter
- Added `solve_discrete_lyapunov`, `solve_cotinuous_lyapunov`, `solve_discrete_are`, and `solve_triangular` to  `__all__`, so there are no more "hidden functions" in `nlinalg`

### Implementation details
`Op` is just a wrapper around `scipy.linalg.solve_discrete_are`, with gradients taken from Kao and Hennequin (2020), https://arxiv.org/pdf/2011.11430.pdf

The use-case is solving matrix quadratic equations. It pops up in state space models (the steady state covariance matrix comes from the solution to such an equation). It can also be used to compute the policy function for Bellman equations with linear-quadratic loss function (also relevant to state space modeling).

### Checklist
+ [ ] Explain motivation and implementation 👆
+ [ ] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [ ] Link relevant issues, preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+ [ ] The commits correspond to [_relevant logical changes_](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes). Note that if they don't, we will [rewrite/rebase/squash the git history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_rewriting_history) before merging.
+ [ ] Are the changes covered by tests and docstrings?
+ [ ] Fill out the short summary sections 👇

## Major / Breaking Changes
- ...

## New features
- ...

## Bugfixes
- ...

## Documentation
- ...

## Maintenance
- ...

### CODE REVIEW

File: pytensor/tensor/slinalg.py [hunk 1]
```diff
@@ -772,7 +764,7 @@ def _direct_solve_discrete_lyapunov(A: "TensorLike", Q: "TensorLike") -> TensorV
 
 def solve_discrete_lyapunov(
     A: "TensorLike", Q: "TensorLike", method: Literal["direct", "bilinear"] = "direct"
-) -> TensorVariable:
+) -> Variable:
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:767]: Can't you add a `typing.cast` on the returned variable?

File: pytensor/tensor/slinalg.py [hunk 2]
```diff
@@ -826,10 +818,93 @@ def solve_continuous_lyapunov(A: "TensorLike", Q: "TensorLike") -> TensorVariabl
...
+
+        return pytensor.graph.basic.Apply(self, [A, B, Q, R], [X])
+
+    def perform(self, node, inputs, output_storage):
+        A, B, Q, R = inputs
+        X = output_storage[0]
+
+        if self.enforce_Q_symmetric:
+            Q = 0.5 * (Q + Q.T)
+        X[0] = scipy.linalg.solve_discrete_are(A, B, Q, R).astype(
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:844]: You should match the dtype of node.outputs[0].type instead. 

Imagine a user changes floatX after creating the node.

File: pytensor/tensor/slinalg.py (hunk 1 above)
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:767]: Which one is ultimately correct?

Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py]: I just changed it back, do you think it's worth wrestling with `typing.cast`? It's not done anywhere else in `slinalg`

**Comment by jessegrabowski** [comment]
Does the `solve = Solve()` [here](https://github.com/pymc-devs/pytensor/blob/9a5deee04e2a4a2553e65a098879fcb255ce3a1b/pytensor/tensor/slinalg.py#L400) do anything? It's immediately shadowed by `def solve` in the next line.

**Comment by ricardoV94** [comment]
> Does the `solve = Solve()` [here](https://github.com/pymc-devs/pytensor/blob/9a5deee04e2a4a2553e65a098879fcb255ce3a1b/pytensor/tensor/slinalg.py#L400) do anything? It's immediately shadowed by `def solve` in the next line.

No. I guess it was left accidentally when the paramterized helper was implemented

File: pytensor/tensor/slinalg.py (hunk 1 above)
Comment (ricardoV94) [reply pytensor/tensor/slinalg.py]: If adding it makes mypy happy yes. A proper solution would be to have `make_node` have a return type hint that specifies what kind of Apply is returned (I have no idea if that's possible atm)

Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py]: A function like this:

```
def(x: TensorVariable) -> TensorVariable:
    return typing.cast(SomeApplyNode(x), TensorVariable)
```

Makes the static checker happy, but add two complaints from mypy:

1. error: Cast target is not a type
2. error: Returning Any from function declared to return "TensorVariable[Any, Any]"

So it seems to be 6 of one, half-dozen of the other. Is there no pure TensorType anywhere in the codebase that should be used as a return type hint for functions? `TensorType` is taken (it's a class, not a Type), and there's `_TensorTypeType`, which is a class, but it's hideous for the name and the leading underscore.

Comment (ricardoV94) [reply pytensor/tensor/slinalg.py]: Aren't you putting the arguments to cast in the wrong order?

Comment (ricardoV94) [reply pytensor/tensor/slinalg.py]: > Other functions in slinalg just don't have return type hints, so I can just do that for now I guess.

You can definitely ignore but would be nice to start improving things eventually.

Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py]: Well that's embarrassing

File: tests/tensor/test_slinalg.py [hunk 3]
```diff
@@ -574,7 +578,7 @@ def test_solve_discrete_lyapunov_via_direct_complex():
     A = rng.normal(size=(N, N)) + rng.normal(size=(N, N)) * 1j
     Q = rng.normal(size=(N, N))
     X = f(A, Q)
-    assert np.allclose(A @ X @ A.conj().T - X + Q, 0.0)
+    utt.assert_allclose(A @ X @ A.conj().T - X + Q, 0.0)
```
Comment (ricardoV94) [inline tests/tensor/test_slinalg.py:581]: Any reason to use this instead of `np.testing.assert_allclose`?

Comment (jessegrabowski) [reply tests/tensor/test_slinalg.py:581]: I copied from `test_kron`, which was immediately above my block of tests and used it. I'll change it.

**Review by ricardoV94 — APPROVED** [review]

**Pull request merged**